GVRf/Extensions/3DCursor/IODevices/io_hand_template/src/main/java/com/sample/hand/template/HandTemplateDevice.java

```diff
@@ -0,0 +1,423 @@
+
+    // create Quaternion objects to reduce GC cycles.
+    private static final Quaternionf cursorRotation = new Quaternionf();
+    private Quaternionf boneRotation = new Quaternionf();
+
+    private Future<GVRTexture> boneTexture;
+
+    /**
+     * Construct the Hand device object
+     *
+     * @param context
+     * @param scene
+     */
+    public HandTemplateDevice(GVRContext context, GVRScene scene) {
+        this.context = context;
+
+        // Create a readback buffer and forward it to the native layer
+        readbackBufferB = ByteBuffer.allocateDirect(DATA_SIZE * BYTE_TO_FLOAT);
+        readbackBufferB.order(ByteOrder.nativeOrder());
+        readbackBuffer = readbackBufferB.asFloatBuffer();
+
+        try {
+            boneTexture = context.loadFutureTexture(new GVRAndroidResource
+                    (context, "cube_texture.png"));
+            Future<GVRTexture> jointTexture = context.loadFutureTexture(new GVRAndroidResource
+                    (context, "cube_texture.png"));
+            rightHand = new IOHand(context);
+            leftHand = new IOHand(context);
+
+            // Uncomment the following lines to add scene objects to joints
+            //setJointSceneObject(rightHand, jointTexture);
+            //setJointSceneObject(leftHand, jointTexture);
+            setPalmSceneObject(rightHand, jointTexture);
+            setPalmSceneObject(leftHand, jointTexture);
+        } catch (IOException e) {
+            Log.e(TAG, "IO Exception", e);
+        }
+        rightDevice = new HandIODevice(RIGHT_HAND_ID + IOFinger.getString(IOFinger.INDEX),
+                scene, "Right Hand");
+        leftDevice = new HandIODevice(LEFT_HAND_ID + IOFinger.getString(IOFinger.INDEX), scene,
+                "Left Hand");
+
+        GVRCameraRig mainCameraRig = scene.getMainCameraRig();
+        mainCameraRig.addChildObject(leftHand.getSceneObject());
+        mainCameraRig.addChildObject(rightHand.getSceneObject());
+
+        /**
+         * It is important that we create a new thread so that the GL and the Main Android threads
+         * are not impacted by the events generated with by the IO device.
+         *
+         * We are intentionally creating a thread in Java since it is easier to maintain. Threads
+         * created in C++ need to be explicitly registered with the JVM. Here is more information
+         * on creating threads in C++ if you wish to do so:
+         * http://android.wooyd.org/JNIExample/
+         */
+        Log.d(TAG, "Creating a new thread");
+        thread = new Thread(threadRunnable);
+        thread.start();
+    }
+
+    void setJointSceneObject(IOHand hand, Future<GVRTexture> jointTexture) {
+        for (int i = 0; i < 5; i++) {
+            IOFinger finger = hand.getIOFinger(i);
+            for (int j = 1; j < 5; j++) {
+                // ignore index tip
+                if (!(i == IOFinger.INDEX && j == IOJoint.TIP)) {
+                    IOJoint joint = finger.getIOJoint(j);
```

Comment:
Don't think it would be an integration step. Would be more of a customization. Not necessarily a TODO. Will add a line comment.